### ISSUE TITLE
OpenSSL does not enforce set signature algorithms

### ISSUE THREAD

**Issue by david-hoze** [openssl/openssl#11438]
Hello, using OpenSSL 1.1.1 (libssl-dev on Ubuntu - 1.1.1-1ubuntu2.1~18.04.5).

I'm trying to disable SHA1 from the Signature Algorithms, I've set OpenSSL's signature algorithms like so:

```c++
SSL_CTX_set1_sigalgs_list(ctx, "ECDSA+SHA256:rsa_pss_rsae_sha256:RSA+SHA256:ECDSA+SHA384:rsa_pss_rsae_sha384:RSA+SHA384:rsa_pss_rsae_sha512:RSA+SHA512");
```

As you can see there is no SHA1 listed there, the function returns success, and indeed the WireShark shows these algorithms:

```
Signature Hash Algorithms (8 algorithms)
    Signature Algorithm: ecdsa_secp256r1_sha256 (0x0403)
    Signature Algorithm: rsa_pss_rsae_sha256 (0x0804)
    Signature Algorithm: rsa_pkcs1_sha256 (0x0401)
    Signature Algorithm: ecdsa_secp384r1_sha384 (0x0503)
    Signature Algorithm: rsa_pss_rsae_sha384 (0x0805)
    Signature Algorithm: rsa_pkcs1_sha384 (0x0501)
    Signature Algorithm: rsa_pss_rsae_sha512 (0x0806)
    Signature Algorithm: rsa_pkcs1_sha512 (0x0601)
```

However, when I connect to https://sha1-intermediate.badssl.com/, OpenSSL does not enforce the signature algorithms I set, and lets the TLS communication go on uninterrupted.

This is the WireShark [pcap file](https://www.dropbox.com/s/ridufox40on6kle/openssl-sha1-disabled.pcapng?dl=0).

Why doesn't OpenSSL enforce the set signature algorithms? I only managed to do so by using [SSL_CTX_get_security_level](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_get_security_level.html) and setting it to 2, but it's not recommended to set it above 1. Also I would like more granular control.

Thanks for your help, David.

**Comment by t8m** [comment]
That is because the SignatureAlgorithms value does not affect the validation of the certificate signatures. It affects only the TLS handshake signatures.

**Comment by kaduk** [comment]
But the `signature_algorithms_cert` extension should affect what's allowed in certificates...though we currently don't generate it or have a way to configure it.

**Comment by kaduk** [comment]
On the other hand, RFC 8446 says """If no "signature_algorithms_cert" extension is present, then the "signature_algorithms" extension also applies to signatures appearing in certificates.""", so one could perhaps argue that we should have those checks.

**Comment by davidben** [comment]
Right, this is the question of whether the X.509 signature algorithms advertisement in TLS is a hint or something binding. I would argue that, given the separation between TLS and X.509, how fussy X.509 verification can be (folks trying for a single "chain" being compatible across different sets of trust anchors, etc.), I think a hint is more prudent. In practice, everyone has just the one, maybe two, certificates to choose from anyway.

I've seen annoying effects coming out of treating it as something binding, such as:
* Servers which are TLS 1.2 intolerant because they have configured a redundant MD2 (!!!) self-signed trust anchor. TLS 1.0/1.1 ignore it. TLS 1.2 breaks if the server tries to enforce this.
* Servers which can sign SHA-2 in ServerKeyExchange just fine, but will break if ClientHello sigalgs omits SHA-1 because they've configured a redundant SHA-1 self-signed trust anchor. (It does not work to reject SHA-1 but still advertise it because an even larger fraction of servers preferentially sign SHA-1 but are perfectly capable of signing SHA-2.)

It would be nice if the ecosystem here were a bit more coherent, but we're not in that world yet. :-/

**Comment by t8m** [comment]
But it is quite clear that the signature algorithms in the self-signed certs should not be considered by the check at all. The self-signature does not provide any security anyway. So I do not think these two examples above are that much relevant.

Are there any real-world cases where signatures in the untrusted chain would be affected by making this binding?

I mean SHA1 signatures in certs are already disallowed by the major web browsers, aren't they? Or is there only CAB restriction on CAs to not produce any?

**Comment by davidben** [comment]
They are relevant if you apply the signature list as a check over the certificates which are actually sent, which is what some TLS implementations do and have caused problems in the above, real-world cases.

Web browsers disallow SHA-1 signatures in the certificate chain *after* path-building. This may vary from what is actually sent. The server may still send SHA-1-signed certificate as self-signed trust anchors. This is pointless but some servers do this. The server may also send a SHA-1-signed intermediate certificate that the browser doesn't use. For instance, that intermediate certificate may be a SHA-1 cross-sign from an older trust anchor to a newer one. Such servers would then be trying to remain compatible with both old clients that still accept SHA-1 and lack the newer trust anchor, and new clients that reject SHA-1 but have the newer trust anchor. The latter would stop at the new trust anchor in path-building and ignore the SHA-1 cross-sign.

Policies applies on the certificate chain after path-building shouldn't be especially troublesome. Then you're just saying you want to configure the accepted algorithms in your X.509 and TLS bits similarly, which is sensible enough. The headache just comes from the discrepancy between signatures actually used in X.509 verification, and signatures observed over the wire by TLS.

**Comment by davidben** [comment]
I realized my comment from yesterday may be confusing for folks not familiar with the TLS handshake. There are two sets of signatures in TLS. The certificates carry signatures, produced by the CA. The TLS handshake itself also (usually) carries a signature (in ServerKeyExchange or CertificateVerify), produced by the TLS server implementation.

For browsers, SHA-1 is out of the CA signatures, but getting SHA-1 out of the TLS server signatures is still [ongoing](https://groups.google.com/a/chromium.org/d/msg/blink-dev/yaJcs4p9LNI/haZWzX-UBwAJ). The two parameter negotiations are tied together in TLS, which is why quirks in taking SHA-1 out of one impact taking SHA-1 out of the other.

**Comment by jvdsn** [comment]
I hate to revive an old issue but I was wondering this is still on the radar of the OpenSSL project. I would be happy to take a stab at adding a check (probably gated by a strict flag of some sort) that verifies the server cert (and client cert if applicable) is consistent with the signature_algorithms sent by the client. On the other hand, if this is not desired at all by the project, perhaps this issue should be closed as wontfix.

RFC 5246 (TLS 1.2) states:
> If the client provided a "signature_algorithms" extension, then all certificates provided by the server MUST be signed by a hash/signature algorithm pair that appears in that extension.

([7.4.2](https://datatracker.ietf.org/doc/html/rfc5246#section-7.4.2).  Server Certificate)

> Any certificates provided by the client MUST be signed using a hash/signature algorithm pair found in supported_signature_algorithms.
> The end-entity certificate provided by the client MUST contain a key that is compatible with certificate_types.  If the key is a signature key, it MUST be usable with some hash/signature algorithm pair in supported_signature_algorithms.

([7.4.4](https://datatracker.ietf.org/doc/html/rfc5246#section-7.4.4).  Certificate Request)

>The certificates MUST be signed using an acceptable hash/signature algorithm pair, as described in [Section 7.4.4](https://datatracker.ietf.org/doc/html/rfc5246#section-7.4.4).  Note that this relaxes the constraints on certificate-signing algorithms found in prior versions of TLS.

([7.4.6](https://datatracker.ietf.org/doc/html/rfc5246#section-7.4.6).  Client Certificate)

Similarly, RFC 8446 (TLS 1.3) states:
> The following rules apply to the certificates sent by the server:
> [...]
> The server's end-entity certificate's public key (and associated restrictions) MUST be compatible with the selected authentication algorithm from the client's "signature_algorithms" extension (currently RSA, ECDSA, or EdDSA).
> The certificate MUST allow the key to be used for signing (i.e., the digitalSignature bit MUST be set if the Key Usage extension is present) with a signature scheme indicated in the client's "signature_algorithms"/"signature_algorithms_cert" extensions (see [Section 4.2.3](https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.3)).

([4.4.2.2](https://datatracker.ietf.org/doc/html/rfc8446#section-4.4.2.2).  Server Certificate Selection)

> In prior versions of TLS, the CertificateRequest message carried a list of signature algorithms and certificate authorities which the server would accept.  In TLS 1.3, the former is expressed by sending the "signature_algorithms" and optionally "signature_algorithms_cert" extensions.  The latter is expressed by sending the "certificate_authorities" extension (see [Section 4.2.4](https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.4)).

([4.3.2](https://datatracker.ietf.org/doc/html/rfc8446#section-4.3.2).  Certificate Request)

> The certificates MUST be signed using an acceptable signature algorithm, as described in [Section 4.3.2](https://datatracker.ietf.org/doc/html/rfc8446#section-4.3.2).  Note that this relaxes the constraints on certificate-signing algorithms found in prior versions of TLS.

([4.4.2.3](https://datatracker.ietf.org/doc/html/rfc8446#section-4.4.2.3).  Client Certificate Selection)

Of course, all these requirements apply to the party that is sending the certificate (commonly the server). I didn't find anything in the RFCs that says the party receiving the certificate must terminate the connection if the other party sends a certificate that was signed using an unsupported algorithm. However, some government standards (e.g., [PKG_TLS_V1.1](https://www.niap-ccevs.org/MMO/PP/PKG_TLS_V1.1.pdf) Test FCS_TLSC_EXT.3:1) do want this behavior to be enforced.  That's the main reason why I'd like to see this added to OpenSSL.

**Comment by nhorman** [comment]
@jvdsn thank you for your query.  Asking questions like the above is in large part how we identify issues that are still of concern to the community.  This issue clearly fell through the cracks due to time constraints/etc, and was marked with the "help wanted" label, identifying it as an issue that is relevant, but one that we need community effort to help resolve.

As a  note, any issue marked as "help wanted" is available for any community member to tackle as they see fit, so your willingness to help is much appreciated.  As you expressed interested in trying to fix this, I'll assign it to you.  Please open a PR and attach it to the development sidebar on the left so that we can review and merge it when you have a proposed solution.

Thanks!

**Comment by martinschmatz** [comment]
@t8m Thanks for consolidating the discussion to here. 

Minor comment in order not to loose information due to consolidation: #26918 requested two features. 

The first requested feature is - I think - equivalent to what's already discussed here: Only accept a connection on client-side if the sigalgs of all public keys in the certificate chain used to verify the authenticity of the server were advertised in the `ClientHello`. 

The second request is more pragmatic, but IMHO more relevant in real life: Only accept a connection on client-side if the sigalg of the public key of the leaf certificate was advertised in the `ClientHello`, AND if the sigalg of the public key in each certificate in the chain used to verify the authenticity has the same or increased security level compared to the previous (= next lower) certificate in the chain.
Might be well that the first aspect is already covered in current version, but the second aspect ('flat or increasing security level') is certainly not yet available.

**Comment by nhorman** [comment]
Just to ask the question here, is this feature not implementable via a custom verification callback via SSL_CTX_set_cert_verify_callback()?  I.e. set a verification callback, and, when the callback is received, add a check to compare the signature algorithm in the leaf certificate to those advertised in the connection by fetching the advertised algs as returned from SSL_get_sigalgs()?

**Comment by martinschmatz** [comment]
@nhorman Yes, from what I can tell without actual implementation test, this should work.

The background for the original question in #26918 was whether in view of what the standard specifies (but seemingly nobody is following tightly) a 'security out of the box' approach should be used, or not. 
Given that after almost one year, no progress was made in this aspect, I conclude that it is 'or not'.

But thank you for pointing out to the community (and myself) how this could be achieved.
